### DISCUSSION TITLE
Recommended way to render single markdown file

### DISCUSSION THREAD

**Discussion by SuibianP** [mkdocs/mkdocs#3765] [Q&A] [answered]
I have a single large document (~500 LoC) that I wish to automatically deploy using MkDocs for readability. The document is not split, for easier collaboration with non-techy contributors over HackMD and since it is not really that huge. I would like some way for MkDocs to automatically render it by turning heading levels into hierarchy. Is there any plugin or builtin functionality to accomplish this?

I have also considered splitting into directory structure to be automatically picked up by MkDocs in the absence of `nav`, but [mdsplit.py](https://github.com/markusstraub/mdsplit)/[mdsplit.cpp](https://github.com/alandefreitas/mdsplit) both have certain issues with some corner cases like some Markdown extension syntax and Unicode headings, and some settings need to be manually synced with `mkdocs.yml` e.g. slugification.

**Comment by SuibianP** [comment]
I assume this would also be useful for quickly whipping up navigable ad hoc documentation sites from repo readme.

**Comment by pawamoy** [comment]
Have you tried anything? If you pick a theme that shows headings in a sidebar, then it's solved: just build your single-page site and deploy it. Check out available themes here: https://github.com/mkdocs/catalog/?tab=readme-ov-file#-theming. Material for MkDocs has such a sidebar for headings. The `toc.integrate` setting could be useful: https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#navigation-integration.

**Reply by SuibianP** [reply]
Thanks for the reply. I did know that, but want each heading rendered in their individual pages (somewhat like `texi2any --split`) instead of anchored globally to save some clutter, which is actually the main reason why HackMD reading mode does not suffice for my use case. Additionally, for MkDocs Material, this puts everything under a `Home` section which does not allow me to make the first heading level tabs (`navigation.tabs`) instead.

I am admittedly not so familiar with MkDocs internals. Please feel free to transfer this thread to `mkdocs-material` if the issue is deemed theme-specific.

Edit: It is also rather hard to distinguish h5 and h6 from paragraph text. Shifting them up by two or three levels in their own page would be far more legible for deeply nested headings.

**Reply by pawamoy** [reply, marked as answer]
I see, thanks for the clarification. I wasn't sure from your original question if you wanted to split onto multiple pages.

Well, I think this could be a plugin. Basically, keep your Markdown document in the root of the repository, and the `docs` folder empty. Then add a `hooks.py` file in the root, that you list under `hooks` in mkdocs.yml.

In this hooks file, write both the `on_files` and `on_nav` hooks. The first one reads your Markdown document, splits it up and move each section into its own temporary file, and create the corresponding virtual MkDocs files. The `on_nav` hook iterates on this collection of virtual files to build the navigation.

To split your document, you could call another tool in a subprocess, or write some Python code. If processing the file line by line is not robust enough (due to its contents), then you could probably implement a Markdown treeprocessor that does the splitting somehow.
